### PR TITLE
PERF: Release unneeded pipeline terms.

### PR DESCRIPTION
**Pull request by ssanderson**
Refcount pipeline terms during execution and release terms once they're
no longer needed.

This dramatically reduces memory usage on large pipelines.

### CODE REVIEW

**Comment by twiecki** [comment]
:+1: :100:

File: zipline/pipeline/graph.py [hunk 1]
```diff
@@ -130,8 +130,38 @@ def initial_refcounts(self):
         refcounts = self.out_degree()
         for t in self.outputs.values():
             refcounts[t] += 1
+
+        for t in initial_workspace:
+            self.decref_dependencies(t, refcounts)
+
         return refcounts
 
+    def decref_dependencies(self, term, refcounts):
```
Comment (llllllllll) [inline zipline/pipeline/graph.py]: shouldn't decref release the data? It seems like this makes it easier to get the refcounts and workspace out of sync.

Comment (ssanderson) [inline zipline/pipeline/graph.py]: The data might not be in the actual workspace if we put `VWAP` in the initial_workspace without `close` and `volume`.

The design here, essentially, is that dependency resolution is the TermGraph's job, and computation is the PipelineEngine's job.  It's the engine's responsibility to actually do something with the terms that the graph identifies as garbage.  This is analogous to the fact that it's the Engine's job to actually compute in the order that the graph determines in `TermGraph.ordered`.

**Comment by twiecki** [comment]
This is awesome. Can you let me know when this hits staging/prod?

**Pull request merged**